test/test_sparse_csr.py

```diff
@@ -575,6 +579,50 @@ def test_csr_matvec(self, device, dtype):
             with self.assertRaisesRegex(RuntimeError, err_msg):
                 csr.matmul(bad_vec)
 
+    @onlyCUDA
+    @unittest.skipIf(not TEST_SCIPY, "SciPy not found")
+    @dtypes(torch.float32, torch.float64, torch.complex64, torch.complex128)
+    def test_block_addmm(self, device, dtype):
+        def run_test(c, a, b, op_b, op_c, *, alpha=None, beta=None):
+            if dtype.is_complex:
+                alpha = random.random() + 0.3j if alpha is None else alpha
+                beta = random.random() + 0.6j if beta is None else beta
+            else:
+                alpha = random.random() if alpha is None else alpha
+                beta = random.random() if beta is None else beta
+
+            if op_b and a.shape == b.shape:
+                b = b.mH
+
+            actual = torch.addmm(c, a, b, alpha=alpha, beta=beta)
+
+            out = torch.empty_like(c if op_c and a.shape == b.shape else c.mH)
+            torch.addmm(c, a, b, alpha=alpha, beta=beta, out=out)
+
+            a_bsr = sp.bsr_matrix(
+                (
+                    a.values().cpu().numpy(),
+                    a.col_indices().cpu().numpy(),
+                    a.crow_indices().cpu().numpy(),
+                ),
+                shape=a.shape,
+            )
+            expected = alpha * (a_bsr * b.cpu().numpy()) + beta * c.cpu().numpy()
+            self.assertEqual(actual, out)
+            self.assertEqual(actual, expected)
+
+        for index_dtype in [torch.int32, torch.int64]:
+            for (m, n, k), block_size, noncontiguous in zip(itertools.product([1, 5], repeat=3), [1, 2, 3], [True, False]):
+                nnz = random.randint(0, m * k)
+                a = self.genSparseCSRTensor((m, k), nnz, dtype=dtype, device=device, index_dtype=index_dtype)
+                a_data = make_tensor((nnz, block_size, block_size), dtype=dtype, device=device)
+                a = torch._sparse_csr_tensor_unsafe(a.crow_indices(), a.col_indices(), a_data, (m * block_size, k * block_size))
```

Comment:
To be precise, this means the CSR constructor will have different semantics from the COO constructor, right?

Because for COO, you'll need to introduce dense dimensions to support higher dimensional values, i.e.

```
>>> i = [[0, 0, 1], [0, 1, 1]]
>>> v = torch.randn(3, 2, 2)
>>> a = torch.sparse_coo_tensor(i, v, (2, 2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: number of dimensions must be sparse_dim (2) + dense_dim (2), but got 2
>>> a = torch.sparse_coo_tensor(i, v, (2, 2, 2, 2))
>>> a.size()
torch.Size([2, 2, 2, 2])
```

I also understand that you're using the private, unsafe constructor. Does this mean you want to change the behavior or arguments passed to the public constructor later on?

In general I'm in favor these semantics. What we want with block sparse layouts is tiling and not additional dense dimensions in the [Hybrid COO sense](https://pytorch.org/docs/master/sparse.html#hybrid-sparse-coo-tensors).

The only issue is that we don't have metadata right now that distinguishes between dimensions of the data used for tiling vs. dimensions exposed to the user as dense dimensions. [The number of dense dimensions is inferred from the dimensionality of the values](https://github.com/pytorch/pytorch/blob/2f099c7555ab41e55ddfd12ba3ed804e52894820/aten/src/ATen/native/sparse/SparseTensor.cpp#L281-L290). We will want dense dimensions soon as well to help represent data points such as embeddings of sentences of different lengths. 

So, for example, let's say you have 4 Tensors of size ```(3, 128), (15, 128), (6, 128), (3, 128)```. Here a CSR Tensor of shape ```(4, 15, 128)```, where the last dimension is dense, is very useful.

Reply:
Yes, the current use of an unsafe CSR constructor is different from the COO. Of course, the public constructor of CSR matrices and conversion functions will need to be modified to be able to create inputs for these new paths.
My plan is to make everything work and tests passing using an unsafe constructor first, keeping it all private. Then discuss & implement constructors with block information separately unlocking the functions for wider use. Documentation will come with the update to constructors.

 > We will want dense dimensions soon as well to help represent data points such as embeddings of sentences of different lengths.
So, for example, let's say you have 4 Tensors of size (3, 128), (15, 128), (6, 128), (3, 128). Here a CSR Tensor of shape (4, 15, 128), where the last dimension is dense, is very useful.

Please put this information into a separate feature request issue.
I know that the most generic representation, that is possible to create in PyTorch, is a CSR matrix of shape `(rows, columns, *dense_shape)`, where rows and columns represent matrix dimensions and `dense_shape` can be shape of any length. However, tensors with this generic representation wouldn't be usable with math libraries like MKL Sparse, cuSPARSE. The libraries expect the `dense_shape` to be strictly either `()` for standard CSR format or `(block_size, block_size)` for standard BSR format.